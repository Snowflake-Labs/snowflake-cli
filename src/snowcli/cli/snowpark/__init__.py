import typer

from snowcli.cli.snowpark.function import app as function_app
from snowcli.cli.snowpark.package import app as package_app
from snowcli.cli.snowpark.procedure import app as procedure_app

app = typer.Typer(
    name="snowpark",
    context_settings={"help_option_names": ["-h", "--help"]},
    help="Manage functions, procedures and Snowpark objects",
)

app.add_typer(function_app)
app.add_typer(package_app)
app.add_typer(procedure_app)
