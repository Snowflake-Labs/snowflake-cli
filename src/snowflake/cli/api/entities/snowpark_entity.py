from snowflake.cli.api.entities.common import EntityBase


class SnowparkEntity(EntityBase):
    pass


class FunctionEntity(SnowparkEntity):
    """
    A single UDF
    """

    pass


class ProcedureEntity(SnowparkEntity):
    """
    A stored procedure
    """

    pass
