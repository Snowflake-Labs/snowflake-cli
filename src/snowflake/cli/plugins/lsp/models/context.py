# coding: utf-8

"""
    LSP Contracts

    No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)

    The version of the OpenAPI document: 1.0.0
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


from __future__ import annotations
import pprint
import re  # noqa: F401
import json

from pydantic import BaseModel, ConfigDict, StrictStr
from typing import Any, ClassVar, Dict, List, Optional
from snowflake.cli.plugins.lsp.models.connection import Connection
from typing import Optional, Set
from typing_extensions import Self

class Context(BaseModel):
    """
    Context
    """ # noqa: E501
    connection: Optional[Connection] = None
    env: Optional[Dict[str, StrictStr]] = None
    project_path: Optional[StrictStr] = None
    __properties: ClassVar[List[str]] = ["connection", "env", "project_path"]

    model_config = ConfigDict(
        populate_by_name=True,
        validate_assignment=True,
        protected_namespaces=(),
    )


    def to_str(self) -> str:
        """Returns the string representation of the model using alias"""
        return pprint.pformat(self.model_dump(by_alias=True))

    def to_json(self) -> str:
        """Returns the JSON representation of the model using alias"""
        # TODO: pydantic v2: use .model_dump_json(by_alias=True, exclude_unset=True) instead
        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, json_str: str) -> Optional[Self]:
        """Create an instance of Context from a JSON string"""
        return cls.from_dict(json.loads(json_str))

    def to_dict(self) -> Dict[str, Any]:
        """Return the dictionary representation of the model using alias.

        This has the following differences from calling pydantic's
        `self.model_dump(by_alias=True)`:

        * `None` is only added to the output dict for nullable fields that
          were set at model initialization. Other fields with value `None`
          are ignored.
        """
        excluded_fields: Set[str] = set([
        ])

        _dict = self.model_dump(
            by_alias=True,
            exclude=excluded_fields,
            exclude_none=True,
        )
        # override the default output from pydantic by calling `to_dict()` of connection
        if self.connection:
            _dict['connection'] = self.connection.to_dict()
        return _dict

    @classmethod
    def from_dict(cls, obj: Optional[Dict[str, Any]]) -> Optional[Self]:
        """Create an instance of Context from a dict"""
        if obj is None:
            return None

        if not isinstance(obj, dict):
            return cls.model_validate(obj)

        _obj = cls.model_validate({
            "connection": Connection.from_dict(obj["connection"]) if obj.get("connection") is not None else None,
            "env": obj.get("env"),
            "project_path": obj.get("project_path")
        })
        return _obj


