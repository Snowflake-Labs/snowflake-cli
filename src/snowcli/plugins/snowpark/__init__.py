from snowcli.plugins.snowpark.commands import app
from snowcli.plugins.snowpark.package.commands import app as package_app
from snowcli.plugins.snowpark.procedure_coverage.commands import (
    app as procedure_coverage_app,
)

app.add_typer(package_app)
app.add_typer(procedure_coverage_app)
