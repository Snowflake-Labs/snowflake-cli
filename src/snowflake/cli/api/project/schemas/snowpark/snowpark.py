from __future__ import annotations

from typing import List

from pydantic import Field
from snowflake.cli.api.project.schemas.snowpark.callable import (
    FunctionSchema,
    ProcedureSchema,
)
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel


class Snowpark(UpdatableModel):
    project_name: str = Field(title="Project identifier")
    stage_name: str = Field(title="Stage in which projectâ€™s artifacts will be stored")
    src: str = Field(title="Folder where your code should be located")
    functions: List[FunctionSchema] | None = Field(
        title="List of functions defined in the project", default=[]
    )
    procedures: List[ProcedureSchema] | None = Field(
        title="List of procedures defined in the project", default=[]
    )
