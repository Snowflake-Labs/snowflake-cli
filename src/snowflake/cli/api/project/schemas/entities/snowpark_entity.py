# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import List, Literal, Optional, Union

from pydantic import Field, field_validator
from snowflake.cli.api.project.schemas.entities.common import (
    EntityModelBase,
    ExternalAccessBaseModel,
)
from snowflake.cli.api.project.schemas.snowpark.argument import Argument
from snowflake.cli.api.project.schemas.updatable_model import DiscriminatorField


class SnowparkEntityModel(EntityModelBase, ExternalAccessBaseModel):
    handler: str = Field(
        title="Function’s or procedure’s implementation of the object inside source module",
        examples=["functions.hello_function"],
    )
    returns: str = Field(
        title="Type of the result"
    )  # TODO: again, consider Literal/Enum
    signature: Union[str, List[Argument]] = Field(
        title="The signature parameter describes consecutive arguments passed to the object"
    )
    runtime: Optional[Union[str, float]] = Field(
        title="Python version to use when executing ", default=None
    )
    imports: Optional[List[str]] = Field(
        title="Stage and path to previously uploaded files you want to import",
        default=[],
    )
    stage: str = Field(title="Stage in which artifacts will be stored")
    artifacts: str = Field(title="Folder where your code should be located")

    @field_validator("runtime")
    @classmethod
    def convert_runtime(cls, runtime_input: Union[str, float]) -> str:
        if isinstance(runtime_input, float):
            return str(runtime_input)
        return runtime_input


class ProcedureEntityModel(SnowparkEntityModel):
    type: Literal["procedure"] = DiscriminatorField()  # noqa: A003
    execute_as_caller: Optional[bool] = Field(
        title="Determine whether the procedure is executed with the privileges of "
        "the owner (you) or with the privileges of the caller",
        default=False,
    )


class FunctionEntityModel(SnowparkEntityModel):
    type: Literal["function"] = DiscriminatorField()  # noqa: A003
