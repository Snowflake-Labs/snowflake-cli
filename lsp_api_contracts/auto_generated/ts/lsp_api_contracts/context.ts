/**
 * LSP Contracts
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { RequestFile } from './models';
import { Connection } from './connection';

export class Context {
    'connection'?: Connection;
    'env'?: { [key: string]: string; };
    'projectPath'?: string;

    static discriminator: string | undefined = undefined;

    static attributeTypeMap: Array<{name: string, baseName: string, type: string}> = [
        {
            "name": "connection",
            "baseName": "connection",
            "type": "Connection"
        },
        {
            "name": "env",
            "baseName": "env",
            "type": "{ [key: string]: string; }"
        },
        {
            "name": "projectPath",
            "baseName": "project_path",
            "type": "string"
        }    ];

    static getAttributeTypeMap() {
        return Context.attributeTypeMap;
    }
}

