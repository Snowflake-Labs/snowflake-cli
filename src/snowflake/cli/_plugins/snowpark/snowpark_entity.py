import functools
from typing import Generic, TypeVar, List

from snowflake.cli._plugins.snowpark.snowpark_entity_model import (
    FunctionEntityModel,
    ProcedureEntityModel,
)
from snowflake.cli._plugins.workspace.context import ActionContext
from snowflake.cli.api.entities.common import EntityBase, get_sql_executor

T = TypeVar("T")


class SnowparkEntity(EntityBase[Generic[T]]):

    @property
    def project_root(self):
        return self._workspace_ctx.project_root

    @property
    def identifier(self):
        return self.model.fqn.sql_identifier

    @property
    def fqn(self):
        return self.model.fqn

    @functools.cached_property
    def _sql_executor(self): # maybe this could be moved to parent class, as it is used in streamlit entity as well
        return get_sql_executor()

    @functools.cached_property
    def _conn(self):
        return self._sql_executor._conn  # noqa

    @property
    def model(self):
        return self._entity_model  # noqa

    def action_bundle(self, action_ctx: ActionContext, *args, **kwargs):
        pass

    def action_deploy(self, action_ctx: ActionContext, *args, **kwargs):
        pass

    def action_drop(self, action_ctx: ActionContext, *args, **kwargs):
        return self._sql_executor.execute_query(self.get_drop_sql())

    def action_describe(self, action_ctx: ActionContext, *args, **kwargs):
        return self._sql_executor.execute_query(self.get_describe_sql())

    def action_execute(self, action_ctx: ActionContext, execution_arguments: List[str] = None, *args, **kwargs):
        return self._sql_executor.execute_query(self.get_execute_sql(execution_arguments))

    def bundle(self, bundle_root):
        pass

    def get_deploy_sql(self):
        pass

    def get_describe_sql(self):
        return f"DESCRIBE {self.identifier}"

    def get_drop_sql(self):
        return f"DROP {self.identifier}"

    def get_execute_sql(self):
        raise NotImplementedError

    def get_usage_grant_sql(self):
        pass

class FunctionEntity(SnowparkEntity[FunctionEntityModel]):
    """
    A single UDF
    """

    def get_execute_sql(self, execution_arguments: List[str] = None, *args, **kwargs):
        if not execution_arguments:
            execution_arguments = []
        return f"SELECT {self.fqn}({', '.join([str(arg) for arg in execution_arguments])})"


class ProcedureEntity(SnowparkEntity[ProcedureEntityModel]):
    """
    A stored procedure
    """

    def get_execute_sql(self, execution_arguments: List[str] = None, ):
        if not execution_arguments:
            execution_arguments = []
        return f"CALL {self.fqn}({', '.join([str(arg) for arg in execution_arguments])})"
