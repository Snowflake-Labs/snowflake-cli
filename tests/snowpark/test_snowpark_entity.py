from pathlib import Path
from unittest import mock

import pytest
import yaml
from markdown_it.rules_inline import entity

from snowflake.cli._plugins.snowpark.snowpark_entity import FunctionEntity
from snowflake.cli._plugins.snowpark.snowpark_entity_model import FunctionEntityModel
from snowflake.cli._plugins.workspace.context import WorkspaceContext, ActionContext
from testing_utils.mock_config import mock_config_key

CONNECTOR = "snowflake.connector.connect"
CONTEXT = ""
EXECUTE_QUERY = "snowflake.cli.api.sql_execution.BaseSqlExecutor.execute_query"


@pytest.fixture
def example_function_workspace(project_directory): #TODO: try to make a common fixture for all entities
    with mock_config_key("enable_native_app_children", True):
        with project_directory("snowpark_functions_v2") as pdir:
            with Path(pdir / "snowflake.yml").open() as definition_file:
                definition = yaml.safe_load(definition_file)
                model = FunctionEntityModel(
                    **definition.get("entities", {}).get("func1")
                )

                workspace_context = WorkspaceContext(
                    console=mock.MagicMock(),
                    project_root=pdir,
                    get_default_role=lambda: "test_role",
                    get_default_warehouse=lambda: "test_warehouse",
                )

                return (
                    FunctionEntity(
                        workspace_ctx=workspace_context, entity_model=model
                    ),
                    ActionContext(
                        get_entity=lambda *args: None,
                    ),
                )


@mock.patch(EXECUTE_QUERY)
def test_action_describe(mock_execute, example_function_workspace):
    entity, action_context = example_function_workspace
    result = entity.action_describe(action_context)

    mock_execute.assert_called_with("DESCRIBE IDENTIFIER('func1')")

@mock.patch(EXECUTE_QUERY)
def test_action_drop(mock_execute, example_function_workspace):
    entity, action_context = example_function_workspace
    result = entity.action_drop(action_context)

    mock_execute.assert_called_with("DROP IDENTIFIER('func1')")

@pytest.mark.parametrize("execution_arguments", [None, ["arg1", "arg2"], ["foo",42,"bar"]])
@mock.patch(EXECUTE_QUERY)
def test_action_execute(mock_execute, execution_arguments, example_function_workspace, snapshot):
    entity, action_context = example_function_workspace
    result = entity.action_execute(action_context, execution_arguments)

    mock_execute.assert_called_with(snapshot)

def test_describe_sql(example_function_workspace):
    entity, _ = example_function_workspace
    assert entity.get_describe_sql() == "DESCRIBE IDENTIFIER('func1')"

def test_drop_sql(example_function_workspace):
    entity, _ = example_function_workspace
    assert entity.get_drop_sql() == "DROP IDENTIFIER('func1')"

@pytest.mark.parametrize("execution_arguments", [None, ["arg1", "arg2"], ["foo",42,"bar"]])
def test_function_get_execute_sql(execution_arguments, example_function_workspace, snapshot):
    entity, _ = example_function_workspace
    assert entity.get_execute_sql(execution_arguments) == snapshot