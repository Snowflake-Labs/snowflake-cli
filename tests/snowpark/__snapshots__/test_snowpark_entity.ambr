# serializer version: 1
# name: test_action_execute[None]
  'SELECT func1()'
# ---
# name: test_action_execute[execution_arguments1]
  'SELECT func1(arg1, arg2)'
# ---
# name: test_action_execute[execution_arguments2]
  'SELECT func1(foo, 42, bar)'
# ---
# name: test_function_get_execute_sql[None]
  'SELECT func1()'
# ---
# name: test_function_get_execute_sql[execution_arguments1]
  'SELECT func1(arg1, arg2)'
# ---
# name: test_function_get_execute_sql[execution_arguments2]
  'SELECT func1(foo, 42, bar)'
# ---
